launchprocess.q

```diff
@@ -0,0 +1,45 @@
+/function which execute system commands from q, and is tailored to deal with the exit codes of lproc.sh:
+/cmd is a string and params are the process parameters which are passed to launch() 
+syscomm:{[params;cmd]
+
+    prevexitcode:first `$system cmd,"; echo $?";
+    $[`0~prevexitcode;
+        .lg.o[`launchprocess;"Successful execution: Starting ",params[`procname]];
+      `1~prevexitcode;
+        .lg.e[`launchprocess;"Error: ",params[`procname]," failed to start"];
+      `2~prevexitcode;
+        .lg.e[`launchprocess;"Error: ",params[`procname]," already running"];
+      .lg.e[`launchprocess;"Unknown error encountered, ",params[`procname]," failed to launch"]]   }
+
+
+
+/function which lets us call launchprocess.sh from inside a TorQ process
+/it takes a dictionary of parameters which will be passed to launchprocess.sh, i.e "-procname rdb1 -proctype rdb" etc.
+
+/ not sure how namespace declaration works     \d .process
+
+launch:{ [params]
+    /exit immediately if process name and type aren't provided
+    if[not 11b~`procname`proctype in key params;0N!
+        .lg.e[`launchprocess;"Process name and type must be provided"];
+        :()];
+    
+    /set default values with .Q.def and string the result for the startline:
+    /for now, let the set of parameters which need to be provided with defaults be:
+    /(procname,proctype),U,localtime
+
+    deflt:`procname`proctype`U`localtime!(`;`;`$getenv[`KDBAPPCONFIG],"/passwords/accesslist.txt";1);
+    params:string each .Q.def[deflt] params; 
+   
+    $[4<count params;
+          [ extras:(-1*(count params)-4)#params; 
+            extraargs:{raze "-",string x[0]," ",string x[1]} each (key extras),'{`$x} each value extras];
```

Comment:
Why cast the values to symbols and then string them?

```suggestion
            extraargs:{"-",string x," ",y}'[key extras;value extras]];
```

this is probably a little cleaner - don't think you should need the `raze`

Reply:
What you suggest there is what I tried initially, but I was getting some issues with the keys being syms and the values strings, i.e the result would be something like ("-";"procname";" ";"w";"d";"b";"1";....) so I ended up needing the raze. In any case, I've reshuffled things since your suggestion to treat all params the same, so everything is much simpler (but I do still have that raze).

Reply:
sorry my suggestion should have been

```suggestion
            extraargs:{"-",string[x]," ",y}'[key extras;value extras]];
```

without the `[]` around `x`, the join will happen first and then `string` will operate on mixed list. this way you should avoid the need to `raze`

Reply:
Ah of course yeah that works, thanks.